### ISSUE TITLE
[BUG] LSP Message (dartls) Invalid line number

### ISSUE THREAD

**Issue by p4-k4** [nvim-flutter/flutter-tools.nvim#403]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

Pasting 60-odd lines of code crashes LSP with `LSP Message (dartls) Invalid line number`.

Restart neovim to continue until next paste (LSP crashes and cannot be restarted).

### Expected Behavior

LSP to remain active without errors.

### Steps To Reproduce

1. Copy/yank around 60 lines of text.
2. Paste into file.

### Environment

```Markdown
NVIM v0.10.1
MacOS

paka  ~/.config/nvim  ♥ 21:16  which flutter 
/opt/homebrew/bin/flutter

[✓] Flutter (Channel master, 3.26.0-1.0.pre.372, on macOS 15.0 24A335 darwin-arm64, locale en-US)
```

**Comment by Mike20403** [comment]
got the same problems , when copy & paste a whole block a code
![image](https://github.com/user-attachments/assets/0af4ef1a-ed30-4c68-b24c-d759022ac852)

**Comment by p4-k4** [comment]
What's your workaround @Mike20403 ? I don't know Lua well at all but trying to look into it.

**Comment by p4-k4** [comment]
Found out that this only occurs when yanking/pasting using a `G` in visual mode to select to bottom.

If you yank/paste from the start of your selection and manually select the last character, no error occurs.

**Comment by Mike20403** [comment]
yep, that's what I'm suffering right now, but I feel a little inconvenient to do so, cause I frequent use ggVG to yank and copy the whole think and paste to another file. Do you know what caused this problem?

**Comment by p4-k4** [comment]
@Mike20403 This seems to occur only during `VG` as you, it's as if it's counting a non-existing, additional line.

Another way I confirmed this was by doing:

1. `ggXXj` (where X = last line number)
2. `y`
3. `ggXXj`(where X = last line number)
4. `p`

This works fine, just as manual selection by character does.

Are you confident with Lua @Mike20403 ? I spent $10 NZD with [`claude.vim`](https://github.com/pasky/claude.vim?tab=readme-ov-file), looking into `flutter-tools.nvim` but there didn't seem to be anything obvious as it began hallucinating. I wondered then if this problem is not related to `flutter-tools.nvim` specifically.

I see you're Vietnamese, my partner is Vietnamese. Xin chào from New Zealand.

**Comment by bytecodevet** [comment]
Hello guys, I also encountered this problem and solved it in the following way:
After saving (BufWritePost), I checked whether dartls was running. If it wasn't, I restarted it (asynchronously).

Here is an example code:

```lua
-- lua/plugins/lsp.lua
-- Dart
lsp.dartls.setup({})

local function reload_dartls_if_inactive()
	local dartls_client = vim.lsp.get_clients({ name = "dartls" })[0]

	vim.defer_fn(function()
		if dartls_client and not dartls_client.is_stopped() then
			return
		end

		if dartls_client and dartls_client.stop then
			dartls_client.stop()
		end

		local lspconfig = require("lspconfig")
		lspconfig.dartls.setup({})
		vim.cmd("edit")
	end, 2000)
end

vim.api.nvim_create_autocmd("BufWritePost", {
	pattern = "*.dart",
	callback = reload_dartls_if_inactive,
})
```

**Comment by p4-k4** [comment]
A decent workaround for now, cheers @bytecodevet

**Comment by bytecodevet** [comment]
Hi guys. The previous code worked quite stable. But when I tried using settings (`dartls.setup({settings=this})`), it didn't work at all, for example, `completeFunctionCalls`, and `closing tags` from `flutter-tools` also didn't work.

After a few hours of digging through the source code of flutter-tools, I finally found a solution.
```lua
-- Dart
-- if you use flutter-tools lspconfig.dartls.setup you don't need to.
local function reload_dartls_if_inactive()
	local dartls_client
	for _, client in ipairs(vim.lsp.get_clients()) do
		if client.name == "dartls" then
			dartls_client = client
			break
		end
	end

	vim.defer_fn(function()
		if dartls_client and not dartls_client.is_stopped() then
			return
		end

		if dartls_client and dartls_client.stop then
			dartls_client.stop()
		end

		require("flutter-tools.lsp").attach() -- <--- this line.
	end, 2000)
end

vim.api.nvim_create_autocmd("BufWritePost", {
	pattern = "*.dart",
	callback = reload_dartls_if_inactive,
})
```

I dug through the code for several hours, and in the end, it was just one line of code to fix it. 😅

**Comment by Mike20403** [comment]
perfect solution, thanks

**Comment by sidlatau** [comment]
Closing the issue, as there is a proposed fix.
